lib/app.rb

```diff
@@ -0,0 +1,81 @@
+class Application
+
+  attr_accessor :frames
+
+  def initialize(io)
+    @io = io
+    @frames = [[],[],[],[],[],[],[],[],[],[]] 
+  end
+
+  def run
+    play_first_nine_frames
+    play_final_frame
+    @io.puts "Your final score is: #{@frames.flatten.sum}"
+  end
+
+  def play_first_nine_frames
+    @frames[0..8].each_with_index do |frame, i|
+      @io.puts "Frame #{i+1}:"
+
+      frame == add_user_scores(frame)
+
+      if i >= 1
+        previous_frame = @frames[i-1]
+        frame << bonus_points?(previous_frame, frame)
+      end
+    end
+  end
+
+  def play_final_frame
+    @io.puts "Frame 10:"
+    @frames[9] == add_user_scores(@frames[9])
+
+    if @frames[9][0] == 10
+      @io.puts "Roll 2:"
+      @frames[9] << get_user_input
+      if @frames[9][1] == 10
+        third_roll
+      end
+    elsif @frames[9].sum == 10
+      third_roll
+    end
+  end
+
+  def bonus_points?(previous_frame, frame)
+    if previous_frame[0] == 10
+      frame.sum
+    elsif previous_frame[0..1].sum == 10
+      frame[0]
+    else 
+      0
+    end
+  end
+
+  def add_user_scores(frame)
```

Comment:
Love this method! Would have been very easy to get trapped putting this logic elsewhere but this both keeps it super readable and eliminates a lot of space for repetition.